File: 2023/13.rb
input = File.read(File.basename(__FILE__).gsub('rb', 'input')).strip.split("\n\n")

input = input.map { _1.split("\n") }

def refrection(pattern, smudge)
  (0...pattern.size - 1).each do |i|
    diff = 0
    (0...pattern.size).each do |j|
      if i + (i - j) + 1 < pattern.size && i + (i - j) + 1 >= 0 && pattern[j] != pattern[i + (i - j) + 1]
        diff += DidYouMean::Levenshtein.distance(pattern[j], pattern[i + (i - j) + 1])
      end
    end

    return i + 1 if diff == smudge
  end
  nil
end

vertical = 0
horizontal = 0
input.each do |pattern|
  ref = refrection(pattern, 0)
  vertical += ref if ref
  pattern = pattern.map(&:chars).transpose.map(&:join)
  ref = refrection(pattern, 0)
  horizontal += ref if ref
end
p vertical * 100 + horizontal

vertical = 0
horizontal = 0
input.each do |pattern|
  ref = refrection(pattern, 2)
  vertical += ref if ref
  pattern = pattern.map(&:chars).transpose.map(&:join)
  ref = refrection(pattern, 2)
  horizontal += ref if ref
end
p vertical * 100 + horizontal

File: 2023/13.py
input = open(f"{__file__.replace('py', 'input')}").read().strip().split("\n\n")


def reflection(pattern, smudge):
    for i in range(len(pattern) - 1):
        diff = 0
        for j in range(len(pattern)):
            if (
                i + (i - j) + 1 in range(len(pattern))
                and pattern[j] != pattern[i + (i - j) + 1]
            ):
                diff += len(
                    [
                        x
                        for x in range(len(pattern[j]))
                        if pattern[j][x] != pattern[i + (i - j) + 1][x]
                    ]
                )
        if diff == smudge:
            return i + 1
    return None


vertical = 0
horizontal = 0
for pattern in input:
    pattern = [x for x in pattern.split("\n")]
    ref = reflection(pattern, 0)
    if ref is not None:
        vertical += ref
    ref = reflection(list(map(list, zip(*pattern))), 0)
    if ref is not None:
        horizontal += ref
print(vertical * 100 + horizontal)

vertical = 0
horizontal = 0
for pattern in input:
    pattern = [x for x in pattern.split("\n")]
    ref = reflection(pattern, 2)
    if ref is not None:
        vertical += ref
    ref = reflection(list(map(list, zip(*pattern))), 2)
    if ref is not None:
        horizontal += ref
print(vertical * 100 + horizontal)
